h/api/models/test/document_test.py

```diff
@@ -434,6 +435,203 @@ def merge_data(self, request):
         return (master, duplicate)
 
 
+class TestUpdateDocumentMetadata(object):
+
+    def test_it_uses_the_target_uri_to_get_the_document(self,
+                                                        annotation,
+                                                        Document,
+                                                        session):
+        document_uri_dicts = [
+            {
+                'uri': 'http://example.com/example_1',
+                'claimant': 'http://example.com/claimant',
+                'type': 'type',
+                'content_type': None,
+            },
+            {
+                'uri': 'http://example.com/example_2',
+                'claimant': 'http://example.com/claimant',
+                'type': 'type',
+                'content_type': None,
+            },
+            {
+                'uri': 'http://example.com/example_3',
+                'claimant': 'http://example.com/claimant',
+                'type': 'type',
+                'content_type': None,
+            },
+        ]
+
+        document.update_document_metadata(session,
+                                          annotation,
+                                          [],
+                                          document_uri_dicts)
+
+        Document.find_or_create_by_uris.assert_called_once_with(
+            session,
+            annotation.target_uri,
+            [
+                'http://example.com/example_1',
+                'http://example.com/example_2',
+                'http://example.com/example_3',
+            ],
+            created=annotation.created,
+            updated=annotation.updated,
+        )
+
+    def test_if_there_are_multiple_documents_it_merges_them_into_one(
+            self,
+            annotation,
+            Document,
+            merge_documents,
+            session):
+        """If it finds more than one document it calls merge_documents()."""
+        Document.find_or_create_by_uris.return_value = mock.Mock(
+            count=mock.Mock(return_value=3))
+
+        document.update_document_metadata(session, annotation, [], [])
+
+        merge_documents.assert_called_once_with(
+            session,
+            Document.find_or_create_by_uris.return_value,
+            updated=annotation.updated)
+
+    def test_it_calls_first(self, annotation, session, Document):
+        """If it finds only one document it calls first()."""
+        Document.find_or_create_by_uris.return_value = mock.Mock(
+            count=mock.Mock(return_value=1))
+
+        document.update_document_metadata(session, annotation, [], [])
+
+        Document.find_or_create_by_uris.return_value\
+            .first.assert_called_once_with()
```

Comment:
<https://github.com/hypothesis/h/pull/3279/commits/a6572ace75149af48d93165d036d7da7b45ba1e9> is another good example of the same problem.

The reason I didn't use the real db though, was because I was thinking that `h.api.models` is our package that uses and wraps the db and sqlalchemy, and `storage` is a module that uses `models`. When one of our modules is using another one of our own module, we should mock out the other module in the tests.

There is definitely a problem here but I wonder if the problem is that the API of `models` isn't very good and that `sqlalchemy` is leaking out of models and being used directly in `storage`.

I wonder if we could improve the API of `models` and its encapsulation of sqlalchemy, so that `storage` tests could mock models without being so brittle?